cpp/src/strings/convert/convert_datetime.cu

```diff
@@ -911,14 +911,9 @@ struct dispatch_from_timestamps_fn {
                        d_timestamps.size(),
                        pfn);
   }
-  template <typename T, std::enable_if_t<not cudf::is_timestamp<T>()>* = nullptr>
-  void operator()(column_device_view const&,
-                  format_item const*,
-                  size_type,
-                  timestamp_units,
-                  const int32_t*,
-                  char* d_chars,
-                  rmm::cuda_stream_view stream) const
+
+  template <typename T, typename... Args>
+  std::enable_if_t<not cudf::is_timestamp<T>(), void> operator()(Args&&...) const
   {
     CUDF_FAIL("Only timestamps type are expected");
   }
```

Comment:
@robertmaynard, should we consider combining the two operators together, via `if constexpr()`?
```suggestion
  template <typename T>
  void operator()(column_device_view const& d_timestamps,
                  format_item const* d_format_items,
                  size_type items_count,
                  timestamp_units units,
                  const int32_t* d_offsets,
                  char* d_chars,
                  rmm::cuda_stream_view stream) const
  {
    if constexpr(not cudf::is_timestamp<T>())
    {
        CUDF_FAIL("Only timestamps type are expected");
    }

    datetime_formatter<T> pfn{d_timestamps, d_format_items, items_count, units, d_offsets, d_chars};
    thrust::for_each_n(rmm::exec_policy(stream),
                       thrust::make_counting_iterator<cudf::size_type>(0),
                       d_timestamps.size(),
                       pfn);
  }
```

Reply:
(I realize this might be peripheral to the task at hand.)

Reply:
The optimizer may be smart enough to remove the code after the if-statement but compiling with debug without optimization may generate alot of code that will not be executed. There are currently 27 cudf types and only 5 timestamp types. I prefer the original if-enable approach.

Reply:
Hmm... I would have thought `if constexpr()` would let the `else` part of the code be removed, regardless of debug/optimization level. These ought to be equivalent, in terms of generated code.
https://stackoverflow.com/questions/54545565/if-constexpr-vs-if-in-light-of-compiler-optimization-and-code-performance

That said, I'm supportive of keeping the code the way it is.

Reply:
Worth an experiment outside of this PR, IMO.

Reply:
I looked at merging these into a single function using constexpr, but didn't like the visual appearance. Since the `if constexpr` approach means that compiler removes the other side of the conditional you still need 'usage' on each variable otherwise you get unused usage warnings. The end result is this is what the constexpr function looks like:
```cpp
  template <typename T>
  void operator()(column_device_view const& d_timestamps,
                  format_item const* d_format_items,
                  size_type items_count,
                  timestamp_units units,
                  const int32_t* d_offsets,
                  char* d_chars,
                  rmm::cuda_stream_view stream) const
  {
    if constexpr(cudf::is_timestamp<T>())
    {
      datetime_formatter<T> pfn{d_timestamps, d_format_items, items_count, units, d_offsets, d_chars};
      thrust::for_each_n(rmm::exec_policy(stream),
                         thrust::make_counting_iterator<cudf::size_type>(0),
                         d_timestamps.size(),
                         pfn);
    } else  {
      (void) d_timestamps;
      (void) d_format_items;
      (void) items_count;
      (void) units;
      (void) d_offsets;
      (void) d_chars;
      (void) stream;
      CUDF_FAIL("Only timestamps type are expected");
    }
  }
  ```
  
But if people aren't annoyed by this, I can look into switching to `if constexpr`

Reply:
No, sir. Let's leave this where it is, for the moment.